mixer/pkg/config/crd/admit.go

```diff
@@ -0,0 +1,387 @@
+	"k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apiserver/pkg/admission"
+	"k8s.io/client-go/kubernetes"
+	admissionClient "k8s.io/client-go/kubernetes/typed/admissionregistration/v1alpha1"
+	"k8s.io/client-go/tools/cache"
+
+	"istio.io/istio/mixer/pkg/config/store"
+)
+
+const (
+	secretServerKey  = "server-key.pem"
+	secretServerCert = "server-cert.pem"
+	secretCACert     = "ca-cert.pem"
+)
+
+// ControllerOptions contains the configuration for the Istio Pilot validation
+// admission controller.
+type ControllerOptions struct {
+	// The name of the resources this controller can admit.
+	ResourceNames []string
+
+	// ExternalAdmissionWebhookName is the name of the
+	// ExternalAdmissionHook which describes he external admission
+	// webhook and resources and operations it applies to.
+	ExternalAdmissionWebhookName string
+
+	// ServiceName is the service name of the webhook.
+	ServiceName string
+
+	// ServiceNamespace is the namespace of the webhook service.
+	ServiceNamespace string
+
+	// ValidateNamespaces is a list of names to validate. Any
+	// namespace not in this list is unconditionally validated as
+	// good. This is useful when multiple validators are running in
+	// the same cluster managing different sets of namespaces
+	// (e.g. shared test clusters).
+	ValidateNamespaces []string
+
+	// // CAbundle is the PEM encoded CA bundle which will be used to
+	// // validate webhook's service certificate.
+	// CABundle []byte
+
+	// SecretName is the name of k8s secret that contains the webhook
+	// server key/cert and corresponding CA cert that signed them. The
+	// server key/cert are used to serve the webhook and the CA cert
+	// is provided to k8s apiserver during admission controller
+	// registration.
+	SecretName string
+
+	// Port where the webhook is served. Per k8s admission
+	// registration requirements this should be 443 unless there is
+	// only a single port for the service.
+	Port int
+
+	// RegistrationDelay controls how long admission registration
+	// occurs after the webhook is started. This is used to avoid
+	// potential races where registration completes and k8s apiserver
+	// invokes the webhook before the HTTP server is started.
+	RegistrationDelay time.Duration
+
+	// Validator defines the actual logic of validating data.
+	Validator store.BackendValidator
+}
+
+// AdmissionController implements the external admission webhook for validation of
+// pilot configuration.
+type AdmissionController struct {
+	client  kubernetes.Interface
+	options ControllerOptions
+}
+
+// GetAPIServerExtensionCACert gets the Kubernetes aggregate apiserver
+// client CA cert used by the "GenericAdmissionWebhook" plugin
+// admission controller.
+//
+// NOTE: this certificate is provided kubernetes. We do not control
```

Comment:
nit: provided `to` K8s or provided `by` k8s?

Reply:
by. added.